src/shared/XLinkDispatcherImpl.c

```diff
@@ -461,10 +480,48 @@ void dispatcherCloseLink(void* fd, int fullClose)
+        // releasePacketFromStream() which will destroy that packet while it is being used by
+        // the app thread -- leading to memory violation/crash. Using the new move-semantics can
+        // help mitigate this.
+        // Also, there is a problem with the ordering of packet get/release throughout the XLink
+        // code. For example, in this while() loop...this thread is assuming that it is the only
+        // thread that has ever gotten a packet. Errant. Imagine the app thread gets 2 packets that
+        // were stored in slots 0 and 1. These two packets are at the "front" of the queue and
+        // marked as gotten/blocked. Now this while loop's thread runs, it getPacketFromStream()
+        // a NEW packet stored in slot 2. And then this loop calls releasePacketFromStream(). Notice
+        // none of these APIs have packet indices or ids. releasePacketFromStream() always
+        // releases/destroys the packet at the "front" of the circular queue. So it destroys the
+        // packet in slot 0 (instead of the packet it just now got from slot 2). That slot 0 packet
+        // is still being used by the app -- leading to memory violation/crash.
+        // BUGBUG Due to the ordering issues, apps should prefer the move semantic XLinkReadMoveData()
+        // to protect the packets they read from unexpected deallocation
+        XLink_sem_wait(&stream->sem);
         while (getPacketFromStream(stream) || stream->blockedPackets) {
             releasePacketFromStream(stream, NULL);
         }
-
+        XLink_sem_post(&stream->sem);
+        // BUGBUG race condition between above/below functions calls where another thread can grab semaphore and
+        // add/remove packets, or operate on stream. Imagine the dispatcher thread processing a XLINK_READ_REQ
+        // event as part of normal event message handling. While at the same time, the app thread
+        // calls `XLinkResetRemoteTimeout()`. The latter can lead to the app thread itself arriving at this function.
+        // Depending on the execution timing of those two threads, it could lead to the app's thread getting the
+        // semaphore first and running the while() loop above. The dispatcher thread is blocked waiting on the semaphore.
+        // Then the app's thread releases the semaphore and the OS pauses it. That causes the dispatcher thread to
+        // be unblocked and it starts to run the function `getPacketFromStream()` and gets memory pointers from the
+        // circular buffer. Meanwhile, the OS unpauses the app's thread so it calls the next function XLinkStreamReset()
+        // which memset(0) the entire stream struct. This will lead to unpredictable results -- sometimes leaked
+        // memory...sometimes app crash.
+        // TODO Two things to help address this bug is to:
+        //   1) use move semantic, this prevents the memset(0) from causing leak/crash
```

Comment:
I think okay in our case as we intend to move completely to move semantic approach in core

Reply:
Even with move and your PR #30 changes, I think there is still a gap between `XLink_sem_post(&stream->sem);` and `XLinkStreamReset(stream);`

Can a semaphore (across OS's) be destroyed while it is still owned?
Then the code be instead something like...

```c++
        ...
        XLink_sem_wait(&stream->sem);
        while (getPacketFromStream(stream) || stream->blockedPackets) {
            releasePacketFromStream(stream, NULL);
        }
        // *** don't release the semaphore
        // XLink_sem_post(&stream->sem);

        XLinkStreamReset(stream);
    }

// then reorder the code inside XLinkStreamReset()
void XLinkStreamReset(streamDesc_t* stream) {
    if(stream == NULL) {
        return;
    }

    // sets all stream fields, including the packets circular buffer to NULL
    // with no check to see if something is open, packet is "blocked", etc.
    memset(stream, 0, sizeof(*stream));
    stream->id = INVALID_STREAM_ID;

    if(XLink_sem_destroy(&stream->sem)) {
        mvLog(MVLOG_DEBUG, "Cannot destroy semaphore\n");
    }
}
```

Reply:
> XLink wrappers for POSIX semaphore functions (refer sem_overview for details)
In description of standard sem_destroy the following can be noted:
"Destroying a semaphore that other processes or threads are currently
blocked on (in sem_wait(3)) produces undefined behavior."
XLink wrappers use thread-safe reference count and destroy the semaphore only in case
if there are no waiters

The XLink wrapper will act differently here.
Otherwise according to Linux spec, destroying a semaphore which is being waited on by another thread is UB.

With XLink wrapper, the above code seems like it would get stuck - sem_wait will increase refcount, and sem_destroy later in `XLinkStreamReset` will block seemingly indefinitely.

What if instead we do something along the lines of:
```
XLink_sem_wait(&stream->sem); 
while (getPacketFromStream(stream) || stream->blockedPackets) { 
    releasePacketFromStream(stream, NULL); 
}
stream->id = INVALID_STREAM_ID;
XLink_sem_post(&stream->sem);
```

And in `getStreamById`
```
streamDesc_t* getStreamById(void* fd, streamId_t id)
{
    xLinkDesc_t* link = getLink(fd);
    XLINK_RET_ERR_IF(link == NULL, NULL);
    int stream;
    for (stream = 0; stream < XLINK_MAX_STREAMS; stream++) {
        if (link->availableStreams[stream].id == id) {
            XLink_sem_wait(&link->availableStreams[stream].sem);
            if(link->availableStreams[stream].id == INVALID_STREAM_ID)
                return NULL;
            return &link->availableStreams[stream];
        }
    }
    return NULL;
}
```

Reply:
Yes and I would tweak that approach sightly. I will push a new commit on this branch aligned with this conversation.

Reply:
Oh... should I squash and force push?

Reply:
Can you bring in latest `develop` changes as well - to be able to remove the comments on `getNextLink...` and have everything up to date.
You may squash and force push.

Also is the latest commit it, or are you still modifying some parts?

Reply:
I have no more coding on this specific issue planned. I can't create a crash in any testing when paired with my `depthai-core` changes that use move. 👍

I have rebased on the newest `develop`, removed the TODO comments (which resolves the merge conflict), and squashed. I'll forcepush after this post.